File: app/src/main/java/com/github/alexgreench/observerpattern/DataRepository.java
package com.github.alexgreench.observerpattern;

import android.annotation.SuppressLint;
import android.os.Handler;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class DataRepository implements Subject{

    private ArrayList<RepositoryObserver> observers;
    private String data;

    private DataRepository() {
        observers = new ArrayList<>();
        getData();
    }

    private static DataRepository INSTANCE = null;

    public static DataRepository getInstance() {
        if(INSTANCE == null) {
            INSTANCE = new DataRepository();
        }
        return INSTANCE;
    }

    private void getData() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                @SuppressLint("SimpleDateFormat")
                SimpleDateFormat sdf = new SimpleDateFormat("HH:mm:ss");
                String currentDateandTime = sdf.format(new Date());
                setData(currentDateandTime);
                getData();
            }
        }, 250);
    }

    private void setData(String data) {
        this.data = data;
        handleEvent();
    }

    @Override
    public void doSubscribe(RepositoryObserver observer) {
        if(!observers.contains(observer)) {
            observers.add(observer);
        }
    }

    @Override
    public void doUnsubscribe(RepositoryObserver observer) {
        observers.remove(observer);
    }

    @Override
    public void handleEvent() {
        for (RepositoryObserver observer: observers) {
            observer.onDataChanged(data);
        }
    }
}

File: app/src/main/java/com/github/alexgreench/observerpattern/DataRepository.kt
package com.github.alexgreench.observerpattern

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Looper
import java.text.SimpleDateFormat
import java.util.*

class DataRepository private constructor() : Subject {

    private val observers: ArrayList<RepositoryObserver> = ArrayList()
    private var data: String? = null

    init {
        getData()
    }

    private fun getData() {
        Handler(Looper.getMainLooper()).postDelayed({
            @SuppressLint("SimpleDateFormat") val sdf = SimpleDateFormat("HH:mm:ss")
            val currentTime = sdf.format(Date())
            setData(currentTime)
            getData()
        }, 250)
    }

    private fun setData(data: String) {
        this.data = data
        handleEvent()
    }

    override fun doSubscribe(observer: RepositoryObserver?) {
        if (!observers.contains(observer)) {
            if (observer != null) {
                observers.add(observer)
            }
        }
    }

    override fun doUnsubscribe(observer: RepositoryObserver?) {
        observers.remove(observer)
    }

    override fun handleEvent() {
        for (observer in observers) {
            observer.onDataChanged(data)
        }
    }

    companion object {
        val instance: DataRepository by lazy {
            DataRepository()
        }
    }
}